app/services/foreman/provision/ssh.rb

```diff
@@ -1,45 +1,66 @@
-require 'timeout'
+require 'open3'
 
 class Foreman::Provision::Ssh
-  attr_reader :template, :uuid, :results, :address, :username, :options
-
-  def initialize(address, username = "root", options = { })
+  attr_reader :template, :uuid, :address, :username, :options
+
+  # options: {
+  #   template: "/path/to/template",
+  #   uuid:     "uuid",
+  #   key_data: "private-key-content",
+  #   password: "string",
+  # }
+  def initialize(address, username = "root", options = {})
     @username = username
     @address  = address
     @template = options.delete(:template) || raise("must provide a template")
     @uuid     = options.delete(:uuid)     || "#{address}-#{username}"
-    @options  = defaults.merge(options)
+    @options  = options
+    @tmp_key_path = create_tmp_key
 
-    initiate_connection!
+    check_auth_options
+    attempt_connection!
   end
 
   def deploy!
-    logger.debug "about to upload #{template} to remote system at #{remote_script}"
-    scp.upload(template, remote_script)
-    logger.debug "about to execute #{command}"
-    @results = ssh.run(command)
-    log_stdout
-    log_stderr
-    success?
+    scp
+
+    logger.debug "about to execute #{main_command}"
+
+    stdout, stderr, status = Open3.capture3(*ssh_cmd(main_command))
+
+    logger.debug(stdout) if stdout.presence
+    logger.error(stderr) if stderr.presence
+
+    status.success?
   end
 
   private
 
-  def success?
-    return true if results.empty?
-    results.map(&:status).compact == [0]
-  end
+  # Attempt to connect to the remote host and run a simple command
+  def attempt_connection!
+    start_time = Time.now.to_i
+
+    logger.debug "Attempting to connect to #{remote_host}"
+    # TODO: Replace password with ***
+    logger.debug ssh_cmd("pwd")
 
-  def log_stdout
-    results.each do |r|
-      r.stdout.split("\n").each { |l| logger.debug l }
+    while (Time.now.to_i - start_time) < Setting[:ssh_timeout].to_i
```

Comment:
Would it make sense to use `ConnectTimeout` as an option? Though I also know booting hosts at some point will refuse because the SSH service isn't running yet so some retry is probably needed. Still, invoking the command fewer times is probably good.